DataStructures/BinarySearchTree/BinarySearchTree.cs

```diff
@@ -0,0 +1,421 @@
+        /// Checks if the specified key is in the BST.
+        /// </summary>
+        /// <param name="key">The key to search for.</param>
+        /// <returns>true if the key is in the BST, false otherwise.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("Key cannot be null!");
+            }
+
+            return Search(root, key) != null;
+        }
+
+        /// <summary>
+        /// Removes a node with the specified key from the BST.
+        /// </summary>
+        /// <param name="key">The key to remove.</param>
+        /// <returns>true if the operation was successful, false otherwise.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        public bool Remove(TKey key)
+        {
+            /* Removing a node from the BST can be split into three cases:
+             * 0. The node to be removed has no children. In this case, the node can just be removed from the tree.
+             * 1. The node to be removed has one child. In this case, the node's child is moved to the node's parent,
+             * then the node is removed from the tree.
+             * 2. The node to be removed has two children. In this case, we must find a suitable node among the children
+             * subtrees to replace the node. This can be done with either the in-order predecessor or the in-order successor.
+             * The in-order predecessor is the largest node in Left subtree, or the largest node that is still smaller then the
+             * current node. The in-order successor is the smallest node in the Right subtree, or the smallest node that is
+             * still larger than the current node. Either way, once this suitable node is found, remove it from the tree (it
+             * should be either a case 1 or 2 node) and replace the node to be removed with this suitable node.
+             * More information: https://en.wikipedia.org/wiki/Binary_search_tree#Deletion
+             */
+
+            if (key == null)
+            {
+                throw new ArgumentNullException("Key cannot be null!");
+            }
+
+            BinarySearchTreeNode<TKey>? node = Search(root, key);
+            if (node == null)
+            {
+                return false;
+            }
+
+            // Case 0: Node has no children
+            if (node.Left == null && node.Right == null)
+            {
+                if (node == root)
+                {
+                    root = null;
+                }
+                else
+                {
+                    if (node.Parent!.Left == node)
+                    {
+                        node.Parent!.Left = null;
+                    }
+                    else if (node.Parent!.Right == node)
```

Comment:
Codacy found an issue: [Add the missing 'else' clause.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277689)